source/structuring.rst

```diff
@@ -14,402 +14,545 @@ functions is better than copying-and-pasting duplicate bits of code
-gets logically replaced with the thing that it points to.  So, to
-refer to the above, we would include::
+.. _canonical-uri:
 
-    { "$ref": "#/definitions/address" }
+Canonical URI
+~~~~~~~~~~~~~
 
-This can be used anywhere a schema is expected. You will always use ``$ref`` as
-the only key in an object: any other keys you put there will be ignored by the
-validator.
+Just like on the web, it's possible to retrieve the same schema from
+multiple URIs. Therefore, it's a good idea for each schema to define
+the preferred URI to use when referencing the schema. This preferred
+URI is called the "canonical URI" and can be set using the ``$id``
+keyword.
 
-The value of ``$ref`` is a URI-reference, and the part after ``#`` sign (the
-"fragment" or "named anchor") is in a format called `JSON Pointer
-<https://tools.ietf.org/html/rfc6901>`__.
+.. draft_specific::
 
-.. note::
-    JSON Pointer aims to serve the same purpose as `XPath
-    <http://www.w3.org/TR/xpath/>`_ from the XML world, but it is much
-    simpler.
+    --Draft 4
+    In Draft 4, ``$id`` is just ``id`` (without the dollar sign).
+
+The value of ``$id`` is a URI-reference that resolves against the
+`retrieval-uri`. The resulting URI is the canonical URI for the
+schema.
 
-If you're using a definition from the same document, the ``$ref`` value begins
-with the pound symbol (``#``). Following that, the slash-separated items traverse
-the keys in the objects in the document. Therefore, in our example
-``"#/definitions/address"`` means:
+Let's assume the URI ``https://example.com/schema/address`` and
+``https://example.com/schema/billing-address`` both identify the
+following schema.
 
-1) go to the root of the document
-2) find the value of the key ``"definitions"``
-3) within that object, find the value of the key ``"address"``
+.. schema_example::
 
-``$ref`` can resolve to a URI that references another file, so if you prefer to
-include your definitions in separate files, you can also do that.  For
-example::
+    {
+      "$id": "/schemas/address",
 
-    { "$ref": "definitions.json#/address" }
+      "type": "object",
+      "properties": {
+        "street_address": { "type": "string" },
+        "city": { "type": "string" },
+        "state": { "type": "string" }
+      },
+      "required": ["street_address", "city", "state"]
+    }
+
+No matter which URI is used to retrieve this schema, the canonical URI
+will be ``https://example.com/schemas/address``, which is the result
+of the ``$id`` URI-reference resolving against the `retrieval-uri`.
+Implementations will use this canonical URI instead of the
+`retrieval-uri` whenever a URI-reference needs to be resolved.
 
-would load the address schema from another file residing alongside
-this one.
+However, using a relative reference when setting a canonical URI can
+be problematic. For example, we couldn't use this schema as a schema
+literal because there would be no `retrieval-uri` and you can't
```

Comment:
```suggestion
be problematic. For example, we couldn't use this schema as a 
anonymous schema because there would be no `retrieval-uri` and you can't
```

After reading your other review comments, I think you just missed this rename.

Reply:
Yes, missed that one.
```suggestion
be problematic. For example, we couldn't use this schema as an
anonymous schema because there would be no `retrieval-uri` and you can't
```